### DISCUSSION TITLE
Connection stays open randomly after closing/crashing consumer

### DISCUSSION THREAD

**Discussion by Matts** [php-amqplib/RabbitMqBundle#647] [General]
Hi, I have no idea how this or why this happends. But on our enterprise application (with 2 consumer connections always open) it sometimes happends that the consumer restarts or loses connection. After this RabbitMQ Manager shows that there are now 4 connections (instead of the intended 2). This keeps duplicating over time.
Only solution we have found. Is to manually log into the management page and kill the connections manually.

Do you recognise this issue or have a fix for it available?
![image](https://user-images.githubusercontent.com/3371705/115194458-c81d7880-a0ed-11eb-92c8-507640e27804.png)

**Comment by Matts** [comment]
Another one of our applications, that uses the Symfony Messenger, does not have this issue. So I don't think this is an architectural IT issue on our side.

**Comment by Matts** [comment]
Also using SimpleBus/rabbitmq-bundle-bridge but thinking this would be an issue on this level seeing as they don't do anything with connection. And it also happends when just using a normal consumer

I am thinking it may have something to do with the sigkill/sigterm

**Comment by MKCG** [comment]
One workaround is to defined a heartbeat in order for the rabbitmq server to proactively close any dead tcp connection : https://www.rabbitmq.com/heartbeats.html

**Comment by mihaileu** [comment]
@Matts is this solving your issue ?

**Comment by Matts** [comment]
I will check this later today. I remember however something saying in the documentation that the heartbeat should be 2x the time of your running consumer. This can possibly cause an issue as our consumers can sometimes be processing for multiple hours. Will reply with results.

**Comment by Matts** [comment]
What I found, was that adding a graceful termination period to my k8s helm config solved the issue partially 
terminationGracePeriodSeconds https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/#hook-handler-execution

Now is just still the issue that the consumer does not detect a closed TCP connection after running for a while. I expect that with heartbeats/keepalive I may be able to solve it.
